x/config/chooser.go

```diff
@@ -0,0 +1,194 @@
+	Choose string   `config:"choose"` // like round-robin / least-pending
+	Peer   string   `config:"peer"`   // implies with: peer
+	Peers  []string `config:"peers"`  // implies with: peers
+	Etc    attributeMap
+}
+
+// Decode captures a configuration on this ChooserConfig struct.
+func (c *ChooserConfig) Decode(into decode.Into) error {
+	var err error
+
+	err = into(&c.Etc)
+	if err != nil {
+		return fmt.Errorf(`could not decode attributes of outbound peer list chooser and updater config: %v`, err)
+	}
+
+	c.With, err = c.Etc.PopString("with")
+	if err != nil {
+		return fmt.Errorf(`could not decode outbound peer list updater config, "with": %v`, err)
+	}
+
+	c.Choose, err = c.Etc.PopString("choose")
+	if err != nil {
+		return fmt.Errorf(`could not decode outbound peer list/chooser config, "choose": %v`, err)
+	}
+
+	c.Peer, err = c.Etc.PopString("peer")
+	if err != nil {
+		return fmt.Errorf(`could not decode outbound "peer" config: %v`, err)
+	}
+
+	_, err = c.Etc.Pop("peers", &c.Peers)
+	if err != nil {
+		return fmt.Errorf(`could not decode outbound "peers" config: %v`, err)
+	}
+
+	return nil
+}
+
+// BuildChooser translates a chooser configuration into a peer chooser, backed
+// by a peer list bound to a peer list updater.
+func (c ChooserConfig) BuildChooser(transport peer.Transport, identify func(string) peer.Identifier, kit *Kit) (peer.Chooser, error) {
+	// Establish a peer selection strategy.
+
+	// Special case for single-peer outbounds.
+	if c.Peer != "" {
+		if c.With != "" {
+			return nil, fmt.Errorf(`cannot combine "peer": %q and "with": %q`, c.Peer, c.With)
+		}
+
+		if c.Choose != "" {
+			return nil, fmt.Errorf(`cannot combine "peer": %q and "choose": %q`, c.Peer, c.Choose)
+		}
+
+		if len(c.Peers) > 0 {
+			return nil, fmt.Errorf(`cannot combine "peer": %q and "peers": %q`, c.Peer, c.Peers)
+		}
+
+		if len(c.Etc) > 0 {
+			return nil, fmt.Errorf(`unrecognized attributes for outbound peer list/chooser config: %v`, c.Etc)
+		}
+
+		return peerbind.NewSingle(identify(c.Peer), transport), nil
+	}
+
+	// All multi-peer lists may combine a peer chooser/list (for sharding or
+	// load-balancing) and updater (a static or dynamic source of peer
+	// addresses).
+
+	// Build a peer list for a peer selection strategy, like round-robin or
+	// least-pending.
+	// Determined by "choose"
+	list, err := c.BuildList(transport, kit)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build a peer list updater, using "peer", "peers", or "with" a custom
+	// peer updater.
+	binder, err := c.BuildBinder(transport, identify, kit)
+	if err != nil {
+		return nil, err
+	}
+
+	return peerbind.Bind(list, binder), nil
+}
+
+// BuildList translates a peer chooser configuration to a peer list / peer
+// chooser, suitable for a peer selection strategy.  The list will be empty
+// until updated with a peer list updater.
+func (c ChooserConfig) BuildList(transport peer.Transport, kit *Kit) (peer.ChooserList, error) {
+	switch c.Choose {
```

Comment:
I think it would be reasonable to remove these special cases and instead register round-robin and least-pending ChooserSpecs on the configurator, just like we will be doing with HTTP and TChannel TransportSpecs, except that we need one default in the absence of a "choose" attribute.